### PR TITLE
snap/snapenv: preserve XDG_RUNTIME_DIR for classic confinement

### PR DESCRIPTION
**Pull request by zyga**
Snaps using classic confinement should not get a custom XDG_RUNTIME_DIR.
This breaks, for example, Wayland and High-DPI scaling support.

Forum: https://forum.snapcraft.io/t/classic-confinement-breaks-high-dpi-support/13868
Signed-off-by: Zygmunt Krynicki <me@zygoon.pl>

### CODE REVIEW

**Comment by mvo5** [comment]
I wonder if this should be targeted for 2.42.1 ?

**Comment by pedronis** [comment]
> I wonder if this should be targeted for 2.42.1 ?

Jamie mentioned potential breakage in the other direction, so we need his input but might need to go a full normal release cycle.

**Comment by zyga** [comment]
@jdstrand I can make an experimental flag for it but I think that without a call for testing it won't be picked up much. We should combine your bullet point ideas 1, 2 and 3 together to be able to comfortably release this. In 2.43 or 2.44 we can change the value of the flag when unset to "enabled" if there is no regression reported.

**Comment by jdstrand** [comment]
> @jdstrand I can make an experimental flag for it but I think that without a call for testing it won't be picked up much. We should combine your bullet point ideas 1, 2 and 3 together to be able to comfortably release this. In 2.43 or 2.44 we can change the value of the flag when unset to "enabled" if there is no regression reported.

If you do add the experimental flag, then ok from me for 2.42. I agree the experimental flag alone isn't enough and encourage as much testing from everyone as possible. I'll leave it to the snapd team what level of testing they're comfortable with to make this default (I just wanted to outline me concerns).

**Comment by zyga** [comment]
@jdstrand in that case I'll make an experimental flag for it. Does `classic-preserves-xdg-runtime-dir` sound good?

**Comment by jdstrand** [comment]
> @jdstrand in that case I'll make an experimental flag for it. Does `classic-preserves-xdg-runtime-dir` sound good?

Sounds fine

**Comment by zyga** [comment]
@jdstrand this is now implemented behind a feature flag.

File: snap/snapenv/snapenv_test.go [hunk 1]
```diff
@@ -108,14 +110,31 @@ func (ts *HTestSuite) TestUser(c *C) {
...
+	})
 
+	// With the classic-preserves-xdg-runtime-dir feature enabled the snap
+	// per-user environment contains no overrides for XDG_RUNTIME_DIR.
+	f := features.ClassicPreservesXdgRuntimeDir
+	c.Assert(ioutil.WriteFile(f.ControlFile(), nil, 0644), IsNil)
+	env = userEnv(mockClassicSnapInfo, "/root")
 	c.Assert(env, DeepEquals, map[string]string{
-		// NOTE HOME Is absent! we no longer override it
+		// NOTE: Both HOME and XDG_RUMTIME_DIR are not defined here.
```
Comment (anonymouse64) [inline snap/snapenv/snapenv_test.go]:
```suggestion
		// NOTE: Both HOME and XDG_RUNTIME_DIR are not defined here.
```
I don't think it's rumtime here (yet)

File: snap/snapenv/snapenv_test.go [hunk 2]
```diff
@@ -108,14 +110,31 @@ func (ts *HTestSuite) TestUser(c *C) {
...
 func (ts *HTestSuite) TestUserForClassicConfinement(c *C) {
+	dirs.SetRootDir(c.MkDir())
+	defer dirs.SetRootDir("/")
+	c.Assert(os.MkdirAll(dirs.FeaturesDir, 0755), IsNil)
+
+	// With the classic-preserves-xdg-runtime-dir feature disabled the snap
+	// per-user environment contains an override for XDG_RUNTIME_DIR.
 	env := userEnv(mockClassicSnapInfo, "/root")
+	c.Assert(env, DeepEquals, map[string]string{
+		// NOTE: Both HOME and XDG_RUMTIME_DIR are not defined here.
```
Comment (anonymouse64) [inline snap/snapenv/snapenv_test.go]:
```suggestion
		// NOTE: Both HOME and XDG_RUNTIME_DIR are not defined here.
```

:thinking: still not rumtime

File: snap/snapenv/snapenv_test.go [hunk 3]
```diff
@@ -217,15 +236,31 @@ func (ts *HTestSuite) TestParallelInstallUser(c *C) {
...
+	})
 
+	// With the classic-preserves-xdg-runtime-dir feature enabled the snap
+	// per-user environment contains no overrides for XDG_RUNTIME_DIR.
+	f := features.ClassicPreservesXdgRuntimeDir
+	c.Assert(ioutil.WriteFile(f.ControlFile(), nil, 0644), IsNil)
+	env = userEnv(&info, "/root")
 	c.Assert(env, DeepEquals, map[string]string{
-		// NOTE HOME Is absent! we no longer override it
+		// NOTE: Both HOME and XDG_RUMTIME_DIR are not defined here.
```
Comment (anonymouse64) [inline snap/snapenv/snapenv_test.go]:
```suggestion
		// NOTE: Both HOME and XDG_RUNTIME_DIR are not defined here.
```

:smile:

File: snap/snapenv/snapenv_test.go (hunk 1 above)
Comment (zyga) [reply snap/snapenv/snapenv_test.go]: Lol :D

**Pull request merged**